src/Native/Unix/System.Security.Cryptography.Native/pal_ssl.c

```diff
@@ -560,46 +319,91 @@ CryptoNative_SslCtxSetCertVerifyCallback(SSL_CTX* ctx, SslCtxSetCertVerifyCallba
-#define SSL_TXT_AllIncludingNull SSL_TXT_ALL SSL_TXT_Separator SSL_TXT_eNULL
-#define SSL_TXT_NotAnon SSL_TXT_Separator SSL_TXT_Exclusion SSL_TXT_aNULL
-
-int32_t CryptoNative_SetEncryptionPolicy(SSL_CTX* ctx, EncryptionPolicy policy)
+void CryptoNative_SetEncryptionPolicy(SSL_CTX* ctx, EncryptionPolicy policy)
 {
-    const char* cipherString = NULL;
-    bool clearSecLevel = false;
-
     switch (policy)
     {
+        case AllowNoEncryption:
+        case NoEncryption:
+            // No minimum security policy, same as OpenSSL 1.0
+            SSL_CTX_set_security_level(ctx, 0);
+            break;
         case RequireEncryption:
-            cipherString = SSL_TXT_ALL SSL_TXT_NotAnon;
+            // get rid of warning
             break;
+    }
+}
 
-        case AllowNoEncryption:
-            cipherString = SSL_TXT_AllIncludingNull;
-            clearSecLevel = true;
-            break;
+int32_t CryptoNative_SetCiphers(SSL_CTX* ctx, const char* cipherList, const char* cipherSuites)
+{
+    int32_t ret = cipherList != NULL || (CryptoNative_Tls13Supported() && cipherSuites != NULL);
 
-        case NoEncryption:
-            cipherString = SSL_TXT_eNULL;
-            clearSecLevel = true;
-            break;
+    // for < TLS 1.3
+    if (cipherList != NULL)
+    {
+        ret &= SSL_CTX_set_cipher_list(ctx, cipherList);
+    }
+
+    // for TLS 1.3
+    if (CryptoNative_Tls13Supported() && cipherSuites != NULL)
+    {
+        ret &= SSL_CTX_set_ciphersuites(ctx, cipherSuites);
     }
 
-    assert(cipherString != NULL);
+    return ret;
+}
+
+const char* CryptoNative_GetOpenSslCipherSuiteName(SSL* ssl, int32_t cipherSuite, int32_t* isTls12OrLower)
+{
+    unsigned char cs[2];
+    const SSL_CIPHER* cipher;
+    const char* ret;
+
+    cs[0] = (cipherSuite >> 8) & 0xFF;
+    cs[1] = cipherSuite & 0xFF;
+    cipher = SSL_CIPHER_find(ssl, cs);
+
+    if (cipher == NULL)
+        return NULL;
+
+    ret = SSL_CIPHER_get_name(cipher);
+
+    if (ret == NULL)
+        return NULL;
+
+    // we should get (NONE) only when cipher is NULL
+    assert(strcmp("(NONE)", ret) != 0);
 
-    if (clearSecLevel)
+    const char* version = SSL_CIPHER_get_version(cipher);
+    assert(version != NULL);
+    assert(strcmp(version, "unknown") != 0);
+
+    // check if tls1.2 or lower
+    // check most common case first
+    if (strncmp("TLSv1", version, 5) == 0)
+    {
+        const char* tlsver = version + 5;
+        // true for TLSv1, TLSv1.0, TLSv1.1, TLS1.2, anything else is assumed to be newer
+        *isTls12OrLower =
+            tlsver[0] == 0 ||
+            (tlsver[0] == '.' && tlsver[1] >= '0' && tlsver[1] <= '2' && tlsver[2] == 0);
+    }
+    else
     {
-        // No minimum security policy, same as OpenSSL 1.0
-        SSL_CTX_set_security_level(ctx, 0);
+        // if we don't know it assume it is new
+        // worst case scenario OpenSSL will ignore it
+        *isTls12OrLower =
+            strncmp("DTLSv", version, 5) == 0 ||
```

Comment:
There could be another version of DTLS for 1.3 (https://datatracker.ietf.org/doc/draft-ietf-tls-dtls13/), so this should be more defensive.